languages/cpp/reference/exercise-concepts/isogram.md

```diff
@@ -0,0 +1,100 @@
+# Isogram
+
+## Task
+
+Determine if a word or phrase is an isogram.
+
+An isogram (also known as a "nonpattern word") is a word or phrase without a repeating letter, however spaces and hyphens are allowed to appear multiple times.
+
+Examples of isograms:
+
+    lumberjacks
+    background
+    downstream
+    six-year-old
+
+The word `isograms`, however, is not an isogram, because the `s` repeats.
+
+
+## Example
+
+`isogram.h`
+```cpp
+#ifndef ISOGRAM_H
+#define ISOGRAM_H
+
+#include <string>
+
+namespace isogram {
+
+bool is_isogram(const std::string& word);
+
+} // namespace isogram
+
+#endif // ISOGRAM_H
+```
+
+`isogram.cpp`
+```cpp
+#include <unordered_set>
+#include "isogram.h"
+
+bool isogram::is_isogram(const std::string& word) {
+  std::unordered_set<char> entries;
+  for (unsigned char c : word) {
```

Comment:
Wouldn't be better if `c` is a `char`? `std::string` is defined as `std::basic_string<char>` I think this approach might cause problems, or am I wrong? 

In any case I think `c` should be a `const` reference.

Reply:
Instead of `char` or `unsigned char`, `auto` should be used.

It is unnecessary to make a reference `char`,  it only adds a level of indirection with the same data size

Reply:
Yes, `auto` is certainly better and you avoid this kind of issues.

Is the reference unecessary? Wouldn't you avoid a copy using it?

Reply:
There is a reason for `unsigned char` in order to avoid undefined behavior in `std::tolower`. 
You should read more about that at [here](https://en.cppreference.com/w/cpp/string/byte/tolower).

Reply:
> Is the reference unecessary? Wouldn't you avoid a copy using it?

Reference is a pointer, pointer is the size of int. Char is smaller than int, so is going to be aligned to int. You got the same memory being copied but you add new level of indirection, accessing the memory via the pointer, which may or may not be recognized by the compiler or processor.

> There is a reason for `unsigned char` in order to avoid undefined behavior in `std::tolower`.
> You should read more about that at [here](https://en.cppreference.com/w/cpp/string/byte/tolower).

This is very good observation, that should be put in common mistakes. It is never easy with C++ and undefined behavior.

Reply:
Yeah, seems fair enough.

Anyway I agree that that observation should be added into common mistakes, I fell for it too. 😅